### PR TITLE
finish basic login

### PR DESCRIPTION
**Pull request by at-toanngvan**

### CODE REVIEW

File: app/controllers/api/v1/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class Api::V1::SessionsController < ApplicationController
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user && user.authenticate(params[:session][:password])
+      user.auth_token
+      render json: user, status: 200
+    else
+      render json: { errors: 'Invalid email or password' }, status: 422
```
Comment (at-kiennguyen) [inline app/controllers/api/v1/sessions_controller.rb]: I18n.

Comment (at-kiennguyen) [inline app/controllers/api/v1/sessions_controller.rb]: status: :unprocessable_entity
http://billpatrianakos.me/blog/2013/10/13/list-of-rails-status-code-symbols/

File: app/controllers/api/v1/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    end
+  end
+
+  def destroy
+    user = User.find_by(email: params[:session][:email].downcase)
+    user.forget
```
Comment (at-kiennguyen) [inline app/controllers/api/v1/sessions_controller.rb]: render json

File: app/controllers/api/v1/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    if user && user.authenticate(params[:session][:password])
+      user.auth_token
+      render json: user, status: 200
+    else
+      render json: { errors: 'Invalid email or password' }, status: 422
+    end
+  end
+
+  def destroy
+    user = User.find_by(email: params[:session][:email].downcase)
```
Comment (at-kiennguyen) [inline app/controllers/api/v1/sessions_controller.rb]: User.find_by(id: params....)

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    BCrypt::Password.create(string, cost: 10)
+  end
+
+  # Returns a random token.
+  def self.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  def auth_token
+    self.auth_token = User.new_token
```
Comment (at-kiennguyen) [inline app/models/user.rb]: new_token

File: app/controllers/api/v1/sessions_controller.rb (hunk 1 above)
Comment (at-toanngvan) [inline app/controllers/api/v1/sessions_controller.rb]: done

Comment (at-toanngvan) [inline app/controllers/api/v1/sessions_controller.rb]: done

File: app/controllers/api/v1/sessions_controller.rb (hunk 2 above)
Comment (at-toanngvan) [inline app/controllers/api/v1/sessions_controller.rb]: done

File: app/controllers/api/v1/sessions_controller.rb (hunk 3 above)
Comment (at-toanngvan) [inline app/controllers/api/v1/sessions_controller.rb]: done

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  # Returns a random token.
+  def self.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  # Remember user
+  def update_auth_token
+    self.auth_token = User.new_token
```
Comment (at-hoangho) [inline app/models/user.rb]: self.auth_token = new_token

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    update_attribute(:auth_token, User.digest(auth_token))
```
Comment (at-hoangho) [inline app/models/user.rb]: update_attribute(:auth_token, digest(auth_token))

File: app/controllers/api/v1/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    user = User.find_by(email: user_params[:email].downcase)
+    if user && user.authenticate(user_params[:password])
+      user.update_auth_token
+      render json: user, status: :ok
+    else
+      render json: { message: I18n.t('controller.session.invalid') }, status: :unauthorized
+    end
+  end
+
+  def destroy
```
Comment (at-hoangho) [inline app/controllers/api/v1/sessions_controller.rb]: Logic fail

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+  def update_auth_token
+    auth_token = new_token
+    update_attribute(:auth_token, digest(auth_token))
+  end
+
+  # Forgets a user.
+  def remove_auth_token
+    update_attribute(:auth_token, nil)
+  end
```
Comment (at-hoangho) [inline app/models/user.rb]: def auth_token
  Encrypt token_digest
end

**Pull request merged**